llmfoundry/data/packing.py

```diff
@@ -260,124 +258,79 @@ def pad_tensor(tensor: torch.Tensor, pad_value: int):
-            '--yaml-path',
-            type=str,
-            required=True,
-            help='Path to the YAML that defines the workload to profile.')
-        parser.add_argument('--num-devices',
-                            type=int,
-                            default=None,
-                            help='How many devices your run will use.')
-        parser.add_argument('--min',
-                            type=float,
-                            required=True,
-                            help='Smallest packing_ratio to test. Must be >=1.')
-        parser.add_argument(
-            '--max',
-            type=float,
-            required=True,
-            help='Largest packing_ratio to test. Must be larger than `min`.')
-        parser.add_argument(
-            '--num-packing-ratios',
-            type=int,
-            default=10,
-            help=
-            'Number of packing_ratio values (spaced between `min` and `max) to try.'
-        )
-
-        args = parser.parse_args()
-
-        if not os.path.isfile(args.yaml_path):
-            raise FileNotFoundError(
-                '`yaml_path` does not correspond to any existing file.')
-        if args.num_devices < 1:
-            raise ValueError('`num_devices` must be a positive integer.')
-        if args.min < 1.0:
-            raise ValueError('`min` must be >=1.0.')
-        if args.max < args.min:
-            raise ValueError('`max` cannot be less than `min`.')
-        if args.num_packing_ratios < 1:
-            raise ValueError('`num_packing_ratios` must be a positive integer.')
-        return args
-
-    def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
-                         device_batch_size: int):
-        if cfg.name == 'text':
-            return build_text_dataloader(cfg, tokenizer, device_batch_size)
-        elif cfg.name == 'text_denoising':
-            return build_text_denoising_dataloader(cfg, tokenizer,
-                                                   device_batch_size)
-        elif cfg.name == 'finetuning':
-            return build_finetuning_dataloader(cfg, tokenizer,
-                                               device_batch_size)
-        else:
-            raise ValueError(
-                f'Not sure how to build dataloader with config: {cfg}')
+def auto_packing_ratio(dataloader_cfg: DictConfig,
+                       tokenizer: PreTrainedTokenizerBase,
+                       device_batch_size: int):
+    """Find a packing ratio that minimizes padding with zero waste.
+
+    Args:
+        dataloader_cfg (DictConfig): The dataloader configuration for profiling.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        A packing ratio that minimizes padding while maintaining zero waste.
+    """
+    min_ratio = 1
+    max_ratio = dataloader_cfg.dataset.max_seq_len / 100
```

Comment:
This assumes ~100 tokens for the "smallest" examples. Would like to get some feedback on whether or not this is reasonable.

future optimization: profile for this as well?

Reply:
Yeah, I agree that this is probably a pretty good heuristic for getting an upper bound on the packing ratio.

If there's an issue, it's probably that with this heuristic, `num_packing_ratios = 20` might not have enough granularity. It might be worth coming up with a less brute force approach that better guarantees that you'll get fine-grained search in the most relevant range of packing ratios.

Off the top of my head, one idea is to start from 1, and double packing ratio until you hit one that is too high (has waste), then next look at half-way between the highest safe packing ratio you've checked and the lowest >0 waste packing ratio you've checked. Do that until the half-way point is rounded into a `raw_batch_size` that you've already checked. Then pick the best packing ratio from the set you've covered.

Reply:
Thanks for the insight! Getting better granularity sounds like a good next step. 
Another idea: If we feel good about the the max heuristic we could also do a binary search that starts between the min and max.

Separately, I think it would be appropriate to do this as a follow up PR since this PR already has a lot of other refactors and new tests. What do you think? @alextrott16 @dakinggg

Reply:
Fine with me if it's a separate PR, but I would advocate for it being a fast follow-up since the 'auto' implementation will potentially change the packing ratio that it comes up with.

Reply:
done